pkg/controller/release/release_controller.go

```diff
@@ -323,46 +324,51 @@ func (c *Controller) syncOneReleaseHandler(key string) error {
 		c.recorder,
 	)
 
-	// This is a 2-round handler: the 1st round schedules the release on a
-	// set of clusters, and the 2nd round creates associated objects and
-	// finalizes release scheduling process. This approach comes for a
-	// reason: we are eliminating a replication lag problem. While running
-	// the system in production, we observed some cases where a few
-	// sequentual updates on the same object experienced apiserver
-	// rejections due to the passed object outdate state. This happened due
-	// to the synchronisation/replication lag in etcd. This approach helps
-	// to eliminate this as a notification will be delivered once all
-	// parties are in sync.
 	if !releaseHasClusters(rel) {
-		_, err = scheduler.ChooseClusters(rel, false)
-	} else {
-		_, err = scheduler.ScheduleRelease(rel)
+		shouldForce := false
+		rel, err = scheduler.ChooseClusters(rel.DeepCopy(), shouldForce)
+		if err != nil {
+			return err
+		}
+		c.recorder.Eventf(
+			rel,
+			corev1.EventTypeNormal,
+			"ClustersSelected",
+			"Set clusters for %q to %v",
+			controller.MetaKey(rel),
+			rel.Annotations[shipper.ReleaseClustersAnnotation],
+		)
 	}
 
-	if err != nil {
+	var scheduleErr error
+	rel, scheduleErr = scheduler.ScheduleRelease(rel.DeepCopy())
+	if scheduleErr != nil {
 		if shippererrors.ShouldBroadcast(err) {
 			c.recorder.Eventf(
-				rel,
+				initialRel,
 				corev1.EventTypeWarning,
 				"FailedReleaseScheduling",
-				err.Error(),
+				scheduleErr.Error(),
 			)
 		}
-
-		reason := reasonForReleaseCondition(err)
+		reason := reasonForReleaseCondition(scheduleErr)
 		condition := releaseutil.NewReleaseCondition(
 			shipper.ReleaseConditionTypeScheduled,
 			corev1.ConditionFalse,
 			reason,
-			err.Error(),
+			scheduleErr.Error(),
 		)
-		releaseutil.SetReleaseCondition(&rel.Status, *condition)
-
-		if _, updateErr := c.clientset.ShipperV1alpha1().Releases(namespace).Update(rel); updateErr != nil {
-			return shippererrors.NewKubeclientUpdateError(rel, updateErr)
+		releaseutil.SetReleaseCondition(&initialRel.Status, *condition)
```

Comment:
It's a bit surprising to me that we refer to `initialRel` anywhere other than the DeepEqual check a bit later in this func. Is this for a special reason?

Reply:
Nevermind, after reading further I see why, things operating on the release return `nil` on error. Makes sense now.